front50-core/src/main/java/com/netflix/spinnaker/front50/model/plugins/PluginInfoService.java

```diff
@@ -114,76 +159,94 @@ public PluginInfo createRelease(@Nonnull String id, @Nonnull PluginInfo.Release
-
-    return existingRelease
-        .map(
-            r -> {
-              pluginInfo.getReleases().remove(r);
-              pluginInfo.getReleases().add(release);
-              cleanupPreferredReleases(pluginInfo, release);
-              validate(pluginInfo);
-              repository.update(pluginInfo.getId(), pluginInfo);
-              return pluginInfo;
-            })
-        .orElseThrow(
-            () ->
-                new NotFoundException(
-                    String.format(
-                        "Plugin %s with release %s version not found. ",
-                        id, release.getVersion())));
+
+    PluginInfo.Release existingRelease =
+        pluginInfo
+            .getReleaseByVersion(release.getVersion())
+            .orElseThrow(
+                () ->
+                    new NotFoundException(
+                        String.format(
+                            "Plugin %s with release %s version not found. ",
+                            id, release.getVersion())));
+
+    pluginInfo.getReleases().remove(existingRelease);
+    pluginInfo.getReleases().add(release);
+
+    if (release.isPreferred()) {
+      preferRelease(pluginInfo, release);
+    }
+
+    return savePluginInfo(pluginInfo);
   }
 
   public PluginInfo deleteRelease(@Nonnull String id, @Nonnull String releaseVersion) {
     PluginInfo pluginInfo = repository.findById(id);
 
-    new ArrayList<>(pluginInfo.getReleases())
-        .forEach(
-            release -> {
-              if (release.getVersion().equals(releaseVersion)) {
-                pluginInfo.getReleases().remove(release);
-              }
-            });
-    repository.update(pluginInfo.getId(), pluginInfo);
+    Optional<PluginInfo.Release> release = pluginInfo.getReleaseByVersion(releaseVersion);
+    release.ifPresent(it -> pluginInfo.getReleases().remove(it));
+
+    savePluginInfo(pluginInfo);
+
     storageService.ifPresent(it -> it.delete(it.getKey(id, releaseVersion)));
     return pluginInfo;
   }
 
+  /** Prefers the given release and sets all other releases to preferred = false; does not save. */
+  private void preferRelease(@Nonnull PluginInfo pluginInfo, @Nullable PluginInfo.Release release) {
+    String preferredVersion =
+        Optional.ofNullable(release).map(PluginInfo.Release::getVersion).orElse(null);
+    Instant now = Instant.now();
+    String user = AuthenticatedRequest.getSpinnakerUser().orElse("anonymous");
+
+    List<PluginInfo.Release> releases = new ArrayList<>(pluginInfo.getReleases());
+    releases.forEach(
+        it -> {
+          boolean wasPreferredRelease = it.isPreferred();
+          boolean isPreferredRelease = it.getVersion().equals(preferredVersion);
+
+          if (isPreferredRelease != wasPreferredRelease) {
+            it.setPreferred(isPreferredRelease);
+            it.setLastModified(now);
+            it.setLastModifiedBy(user);
+            pluginInfo.getReleases().remove(it);
```

Comment:
is removing and re-adding elements necessary for them to be dirty?  I remember having to do something like this with hibernate back in the day.